client/allocrunner/tasklifecycle/coordinator.go

```diff
@@ -0,0 +1,415 @@
+	}
+
+	for lifecycle := range c.tasksByLifecycle {
+		c.gates[lifecycle] = NewGate(shutdownCh)
+	}
+
+	c.enterStateLocked(coordinatorStateInit)
+	return c
+}
+
+// Restore is used to set the Coordinator FSM to the correct state when an
+// alloc is restored. Must be called before the allocrunner is running.
+func (c *Coordinator) Restore(states map[string]*structs.TaskState) {
+	// Skip the "init" state when restoring since the tasks were likely already
+	// running, causing the Coordinator to be stuck waiting for them to be
+	// "pending".
+	c.currentStateLock.Lock()
+	c.enterStateLocked(coordinatorStatePrestart)
+	c.currentStateLock.Unlock()
+
+	c.TaskStateUpdated(states)
+}
+
+// StartConditionForTask returns a channel that is unblocked when the task is
+// allowed to run.
+func (c *Coordinator) StartConditionForTask(task *structs.Task) <-chan struct{} {
+	lifecycle := taskLifecycleStage(task)
+	return c.gates[lifecycle].WaitCh()
+}
+
+// TaskStateUpdated notifies that a task state has changed. This may cause the
+// Coordinator to transition to another state.
+func (c *Coordinator) TaskStateUpdated(states map[string]*structs.TaskState) {
+	c.currentStateLock.Lock()
+	defer c.currentStateLock.Unlock()
+
+	// We may be able to move directly through some states (for example, when
+	// an alloc doesn't have any prestart task we can skip the prestart state),
+	// so loop until we stabilize.
+	// This is also important when restoring an alloc since we need to find the
+	// state where FSM was last positioned.
+	for {
+		nextState := c.nextStateLocked(states)
+		if nextState == c.currentState {
+			return
+		}
+
+		c.enterStateLocked(nextState)
+	}
+}
+
+// nextStateLocked returns the state the FSM should transition to given its
+// current internal state and the received states of the tasks.
+// The currentStateLock must be held before calling this method.
+func (c *Coordinator) nextStateLocked(states map[string]*structs.TaskState) coordinatorState {
+	switch c.currentState {
+	case coordinatorStateInit:
+		if !c.isInitDone(states) {
+			return coordinatorStateInit
+		}
+		return coordinatorStatePrestart
+
+	case coordinatorStatePrestart:
+		if !c.isPrestartDone(states) {
+			return coordinatorStatePrestart
+		}
+		return coordinatorStateMain
+
+	case coordinatorStateMain:
+		if !c.isMainDone(states) {
+			return coordinatorStateMain
+		}
+		return coordinatorStatePoststart
+
+	case coordinatorStatePoststart:
+		if !c.isPoststartDone(states) {
+			return coordinatorStatePoststart
+		}
+		return coordinatorStateWaitMain
+
+	case coordinatorStateWaitMain:
+		if !c.isWaitMainDone(states) {
+			return coordinatorStateWaitMain
+		}
+		return coordinatorStatePoststop
+
+	case coordinatorStatePoststop:
+		return coordinatorStatePoststop
+	}
+
+	// If the code reaches here it's a programming error, since the switch
+	// statement should cover all possible states and return the next state.
+	c.logger.Warn(fmt.Sprintf("unexpected state %s", c.currentState))
```

Comment:
```suggestion
	c.logger.Warn("unexpected state", "state", c.currentState)
```

Reply:
Would this be a good candidate for a `panic` as well, like in https://github.com/hashicorp/nomad/pull/14009#discussion_r943760369? 🤔

Reply:
Oh! Good call! I think so: The intention seems like the other place: if the method in question doesn't have a case for every possible value, then we goofed.

`<insert unhelpful grumbling about type systems that let you enforce such constraints here>`